tests/unit/C/common/test_circular_buffer.cpp

```diff
@@ -0,0 +1,257 @@
+    //First ReadingSet
+    vector<Reading *> *readings1 = new vector<Reading *>;
+    long dpVal1 = 30;
+    DatapointValue dpv1(dpVal1);
+    readings1->emplace_back(new Reading("R1", new Datapoint("DP1", dpv1)));
+    ReadingSet* rs1 = new  ReadingSet(readings1);
+    buffer.insert(rs1);
+    delete readings1;
+    ASSERT_EQ(buffer.extract(false).size(),1);
+
+    //Second ReadingSet
+    long dpVal2 = 50;
+    DatapointValue dpv2(dpVal2);
+    vector<Reading *> *readings2 = new vector<Reading *>;
+    readings2->emplace_back(new Reading("R2", new Datapoint("DP2", dpv2)));
+    ReadingSet* rs2 = new  ReadingSet(readings2);
+    buffer.insert(rs2);
+    delete readings2;
+    ASSERT_EQ(buffer.extract(false).size(),2);
+
+    //Third ReadingSet
+    long dpVal3 = 40;
+    DatapointValue dpv3(dpVal3);
+    vector<Reading *> *readings3 = new vector<Reading *>;
+    readings3->emplace_back(new Reading("R3", new Datapoint("DP3", dpv3)));
+    ReadingSet* rs3 = new  ReadingSet(readings3);
+    buffer.insert(rs3);
+    delete readings3;
+    ASSERT_EQ(buffer.extract(false).size(),3);
+
+    //Fourth ReadingSet
+    long dpVal4 = 45;
+    DatapointValue dpv4(dpVal4);
+    vector<Reading *> *readings4 = new vector<Reading *>;
+    readings4->emplace_back(new Reading("R4", new Datapoint("DP4", dpv4)));
+    ReadingSet* rs4 = new  ReadingSet(readings4);
+    buffer.insert(rs4);
+    delete readings4;
+    ASSERT_EQ(buffer.extract(false).size(),4);
+
+    //Fifth ReadingSet
+    long dpVal5 = 86;
+    DatapointValue dpv5(dpVal5);
+    vector<Reading *> *readings5 = new vector<Reading *>;
+    readings5->emplace_back(new Reading("R5", new Datapoint("DP5", dpv5)));
+    ReadingSet* rs5 = new  ReadingSet(readings5);
+    buffer.insert(rs5);
+    delete readings5;
+    ASSERT_EQ(buffer.extract(false).size(),5);
+
+    //Sixth ReadingSet
+    long dpVal6 = 75;
+    DatapointValue dpv6(dpVal6);
+    vector<Reading *> *readings6 = new vector<Reading *>;
+    readings6->emplace_back(new Reading("R6", new Datapoint("DP6", dpv6)));
+    ReadingSet* rs6 = new  ReadingSet(readings6);
+    buffer.insert(rs6);
+    delete readings6;
+    ASSERT_EQ(buffer.extract(false).size(),5); // Buffer size can't exceed the default MaxLimit
+}
+
+TEST(TESTCircularBuffer, TestExtactFromEmptyBuffer)
+{
+	ReadingSetCircularBuffer buffer;
+    ASSERT_EQ(buffer.extract(false).size(),0); // Buffer size zero in case of extract from an empty buffer
+
+}
+
+TEST(TESTCircularBuffer, TestHeadAndTailMarkerAdjustment)
+{
+	ReadingSetCircularBuffer buffer(3);
+    //First ReadingSet
+    vector<Reading *> *readings1 = new vector<Reading *>;
+    long dpVal1 = 30;
+    DatapointValue dpv1(dpVal1);
+    readings1->emplace_back(new Reading("R1", new Datapoint("DP1", dpv1)));
+    ReadingSet* rs1 = new  ReadingSet(readings1);
+    buffer.insert(rs1);
+    delete readings1;
+    std::vector<std::shared_ptr<ReadingSet>> buff1 = buffer.extract();
+    ASSERT_EQ(buff1.size(),1);
```

Comment:
May be also compare the reading returned in buff1 with readings1, fieldwise, may be by using toJSON() and doing string comparison.